ProjectName/Sources/Data/NetworkAPI/Core/NetworkAPIProtocol.swift

```diff
@@ -0,0 +1,46 @@
+//
+//  NetworkAPIProtocol.swift
+//
+
+import RxSwift
+import Alamofire
+import RxAlamofire
+
+protocol NetworkAPIProtocol {
+
+    func performRequest<T: Decodable>(_ configuration: RequestConfiguration, for type: T.Type) -> Single<T>
+}
+
+extension NetworkAPIProtocol {
+
+    func request<T: Decodable>(
+        session: Session,
+        configuration: RequestConfiguration,
+        decoder: JSONDecoder
+    ) -> Single<T> {
+        return session.rx.request(
+            configuration.method,
+            configuration.url,
+            parameters: configuration.parameters,
+            encoding: configuration.encoding,
+            headers: configuration.headers,
+            interceptor: configuration.interceptor
+        )
+        .responseData()
+        .observe(on: ConcurrentDispatchQueueScheduler(qos: .background))
+        .flatMap { _, data -> Observable<T> in
+            Observable.create { observer in
+                do {
+                    let decodable = try decoder.decode(T.self, from: data)
+                    observer.on(.next(decodable))
+                } catch {
+                    observer.on(.error(error))
+                }
+                observer.on(.completed)
+                return Disposables.create()
+            }
+        }
+        .observe(on: MainScheduler.instance)
```

Comment:
Can you give more insight into this, why we block the UI thread when calling an API?

This `.observe(on: MainScheduler.instance)` is not flexible to me because sometimes we might call another API after the first call.

`firstAPICall.flatMap { secondAPICall() } .subscribe()`

Reply:
should we bring the observe part into the usecase? this way we can control when should we observe (whether it is a chain request or not). What do you guys think? 🤔